# Naive Bayes Classification Model

# Bank Customer Churn Prediction: An End-to-End Machine Learning Experimentation Pipeline

## Project Overview

Customer churn prediction is one of the most important real-world applications of Machine Learning in the banking and financial sector. Retaining existing customers is significantly more cost-effective than acquiring new ones, making churn analysis a critical business problem for modern organizations.

This project focuses on building a complete end-to-end Machine Learning experimentation pipeline for predicting customer churn using banking customer data. Rather than training a single model and reporting accuracy, this project emphasizes analytical thinking, preprocessing decisions, model evaluation, experimentation, and understanding the impact of different Machine Learning techniques on predictive performance.

---

## Dataset

The dataset used in this project is the Bank Customer Churn Prediction dataset, which contains customer demographic information, banking activity, and churn status used for classification analysis.

Dataset Source:  
[Kaggle - Bank Customer Churn Prediction Dataset](https://www.kaggle.com/datasets/shantanudhakadd/bank-customer-churn-prediction)

---

## The project includes:

- Data preprocessing and feature engineering
- Outlier analysis and business-driven preprocessing decisions
- Categorical feature encoding
- Splitting the dataset into the Train set and Test set
- Feature scaling experiments
- Class imbalance handling techniques
- Model training and evaluation
- Cross-validation for performance stability analysis
- Comparative experimentation across multiple Machine Learning models

A major focus of this project is understanding how preprocessing choices affect model behavior and performance in imbalanced classification problems. Different strategies such as Standard Scaling, Robust Scaling, class weighting, and imbalance handling techniques were explored and critically evaluated rather than applied blindly.

The experiments demonstrate the importance of looking beyond accuracy when evaluating classification models. Metrics such as Precision, Recall, F1-Score, ROC-AUC, and Confusion Matrix analysis were used to assess model effectiveness, particularly for detecting churned customers.

This project was designed not only to build predictive models, but also to develop a deeper understanding of practical Machine Learning workflows used in real-world industry environments.

---

## Project Objectives

- Predict whether a customer is likely to churn
- Analyze the impact of preprocessing techniques on model performance
- Evaluate multiple classification algorithms
- Investigate class imbalance handling methods
- Compare model performance using multiple evaluation metrics
- Build a reproducible and professional Machine Learning experimentation workflow

---

## Technologies and Libraries

- Python
- NumPy
- Pandas
- Scikit-learn
- Imbalanced-learn

---

## Machine Learning Concepts Explored

- Classification
- Logistic Regression
- K-Nearest Neighbors (KNN)
- Support Vector Machine SVM
- Naive Bayes
- Decision Trees
- Random Forest
- One-Hot Encoding
- Feature Scaling
- StandardScaler vs RobustScaler
- Outlier Analysis
- Class Imbalance Handling
- Cross Validation
- Model Evaluation Metrics
- ROC-AUC Analysis


# **Importing the Libraries**

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# **Data preprocessing**

## Data Cleaning Check Template
This template is designed to quickly assess the quality of any dataset before building machine learning models or performing analysis.

It provides a structured overview of the dataset by checking for common data issues such as:

- Missing values

- Duplicate rows

- Incorrect data types

- Outliers

- Distribution of numerical features

- Categorical feature consistency

**What This Template Does**

- Displays basic dataset information (shape, data types)

- Identifies missing values and duplicates

- Summarizes numerical and categorical features

- Detects potential outliers using the IQR method

- Highlights columns with low unique values for quick inspection

How to Use

1. Load your dataset using Pandas  

2. Call the function:

In [2]:
def data_quality_report(df):

    print("DATA QUALITY REPORT")
    
    # Print a separator line for better readability
    
    print("=" * 50)
    print("BASIC INFO")
    print("=" * 50)
    
    # Show general information about the dataset (columns, data types, non-null values)
    print(df.info())
    
    # Show number of rows and columns
    print("\n" + "=" * 50)
    print("SHAPE OF DATA")
    print("=" * 50)
    print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
    
    # Check for missing (null) values in each column
    print("\n" + "=" * 50)
    print("MISSING VALUES")
    print("=" * 50)
    missing = df.isnull().sum()
    
    # Display only columns that have missing values
    print(missing[missing > 0])
    
    # Check for duplicate rows
    print("\n" + "=" * 50)
    print("DUPLICATES")
    print("=" * 50)
    print(f"Duplicate rows: {df.duplicated().sum()}")
    
    # Display data types of each column
    print("\n" + "=" * 50)
    print("DATA TYPES")
    print("=" * 50)
    print(df.dtypes)
    
    # Summary statistics for numerical columns (mean, std, min, max, etc.)
    print("\n" + "=" * 50)
    print("NUMERICAL SUMMARY")
    print("=" * 50)
    print(df.describe())
    
    # Summary for categorical (object) columns
    print("\n" + "=" * 50)
    print("CATEGORICAL SUMMARY")
    print("=" * 50)
    print(df.describe(include=['object']))
    
    # Show unique values for columns with low number of distinct values
    # Useful for detecting categories, errors, or inconsistencies
    print("\n" + "=" * 50)
    print("UNIQUE VALUES (LOW CARDINALITY)")
    print("=" * 50)
    for col in df.columns:
        if df[col].nunique() < 10:  # Only show columns with few unique values
            print(f"{col}: {df[col].unique()}")
            
    # correlation
    print("\n" + "=" * 50)
    print("CORRELATION MATRIX")
    print("=" * 50)
    print(df.corr(numeric_only=True))
    
    # Detect outliers using the IQR (Interquartile Range) method
    print("\n" + "=" * 50)
    print("OUTLIERS CHECK (IQR METHOD)")
    print("=" * 50)
    
    # Loop through only numerical columns
    for col in df.select_dtypes(include=np.number).columns:
        Q1 = df[col].quantile(0.25)  # 25th percentile
        Q3 = df[col].quantile(0.75)  # 75th percentile
        IQR = Q3 - Q1  # Interquartile range
        
        # Count rows that fall outside the normal range
        outliers = df[(df[col] < Q1 - 1.5 * IQR) | (df[col] > Q3 + 1.5 * IQR)]
        print(f"{col}: {len(outliers)} outliers")

# **Load dataset**
Apply Data Cleaning Check Template

In [ ]:
dataset = pd.read_csv("Churn_Modelling.csv")
data_quality_report(dataset)

DATA QUALITY REPORT
BASIC INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB
None

SHAPE OF DATA
Rows: 10000, Columns: 14

MISSING VALUES
Se

# **Outlier Analysis**
Outlier detection was performed using the IQR method on the numerical features.

### Features with Detected Outliers

- `CreditScore`
- `Age`
- `NumOfProducts`
- `Exited`

### Decision About Outliers

The detected outliers were intentionally retained instead of being removed because these values represent realistic customer behavior rather than data entry errors.

#### CreditScore
Extreme credit scores are valid in banking datasets and may contain important information about customer financial behavior and churn patterns. Removing these observations could reduce the model’s ability to learn from high-risk or low-risk customers.

#### Age
Older customers were detected as statistical outliers due to the IQR method. However, these customers are valid observations and may have significantly different churn behavior compared to younger customers. Therefore, they were preserved.

#### NumOfProducts
Customers with a high number of banking products were identified as outliers. These cases are meaningful from a business perspective because product usage can strongly influence customer retention and churn.

#### Exited
The `Exited` feature represents the target variable of the classification problem, where:

- `0` = Customer stayed
- `1` = Customer churned

Although the IQR method detected many values as outliers, this result is not meaningful because `Exited` is a binary categorical variable rather than a continuous numerical feature.

Outlier detection methods such as IQR are designed for continuous numerical distributions and should not be applied to classification labels.

Therefore, no outlier handling was performed on the `Exited` column, and all observations were retained.

### Additional Notes
- Identifier columns such as:
  - `RowNumber`
  - `CustomerId`

  were removed from the dataset because they do not provide predictive value.

- `Surname`

The `Surname` column was removed from the dataset because it represents a high cardinality categorical feature with a large number of unique customer names.

Customer surnames are unlikely to have a direct causal relationship with churn behavior and may introduce noise into the model.

Additionally, encoding this feature could create unnecessary dimensionality or artificial relationships between categories without providing meaningful predictive value.

Therefore, the `Surname` feature was excluded from the modeling process.

### Conclusion

The outliers were retained to preserve potentially valuable business information and avoid removing legitimate customer behavior patterns. Tree-based models such as Random Forest are generally robust to outliers, making aggressive outlier removal unnecessary for this project.

# **Select Independent Variable (X) and Dependent Variable (y)**

Note: The concepts of Independent and Dependent variables were discussed earlier.

If you would like to review or gain a clearer understanding, please refer to this section: [Go to Independent variable VS Dependent variable Section](https://github.com/Hazem1695/ml-concept-briefs)

In [4]:
# Separate independent variable(features (X)) by selecting all rows and all columns except the last one
X = dataset.iloc[:, 3:-1].values
# Separate the dependent variable (Target(y)) by selecting the last column
y = dataset.iloc[:, -1].values

In [5]:
print(X[0:10, :])

[[619 'France' 'Female' 42 2 0.0 1 1 1 101348.88]
 [608 'Spain' 'Female' 41 1 83807.86 1 0 1 112542.58]
 [502 'France' 'Female' 42 8 159660.8 3 1 0 113931.57]
 [699 'France' 'Female' 39 1 0.0 2 0 0 93826.63]
 [850 'Spain' 'Female' 43 2 125510.82 1 1 1 79084.1]
 [645 'Spain' 'Male' 44 8 113755.78 2 1 0 149756.71]
 [822 'France' 'Male' 50 7 0.0 2 1 1 10062.8]
 [376 'Germany' 'Female' 29 4 115046.74 4 1 0 119346.88]
 [501 'France' 'Male' 44 4 142051.07 2 0 1 74940.5]
 [684 'France' 'Male' 27 2 134603.88 1 1 1 71725.73]]


# **Encoding categorical data**

## Encoding the Indpendent Variable Using One Hot Encoding

Note: I explained One-Hot Encoding earlier.

If you missed it or want a deeper understanding of how it works, you can find it here: [Go to One-Hot Encoding Section](https://github.com/Hazem1695/ml-concept-briefs)

In [6]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
# One-hot encode the 'color' column (index 2) while keeping other features
# Initialize ColumnTransformer to handle categorical data
# [2] targets the 'color' column (index 2)
ct = ColumnTransformer(transformers = [('encoder', OneHotEncoder(sparse_output=False, drop='first'),[1,2])], remainder = 'passthrough')
# Execute transformation and convert to NumPy ndarray for model training
X = np.array(ct.fit_transform(X))
#

### Categorical Feature Encoding

Categorical variables were transformed into numerical representations using One-Hot Encoding.

#### Geography
The `Geography` feature was encoded using One-Hot Encoding. Since the feature contains three categories (`France`, `Spain`, and `Germany`), two binary columns were created while one category was automatically dropped to avoid the dummy variable trap and multicollinearity.

Encoding representation:

| Germany | Spain | Geography |
|---|---|---|
| 0 | 0 | France |
| 0 | 1 | Spain |
| 1 | 0 | Germany |

#### Gender
The `Gender` feature was also transformed using One-Hot Encoding. Since it contains only two categories (`Female` and `Male`), a single binary column was generated after dropping one category.

Encoding representation:

| Gender |
|---|
| 0 → Female |
| 1 → Male |

These transformations convert categorical variables into numerical representations suitable for machine learning algorithms.

# **Splitting the dataset into the Train set and Test set**

In [7]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

In [8]:
print(X_train[0:10, :])
#france  0.0 0.0
#spain   0.0 1.0
#germany 1.0 0.0
#female  0.0
#male    1.0

[[1.0 0.0 0.0 579 39 5 117833.3 3 0 0 5831.0]
 [0.0 0.0 0.0 750 32 5 0.0 2 1 0 95611.47]
 [0.0 1.0 0.0 729 34 9 53299.96 2 1 1 42855.97]
 [0.0 1.0 1.0 689 38 5 75075.14 1 1 1 8651.92]
 [0.0 0.0 1.0 605 52 7 0.0 2 1 1 173952.5]
 [0.0 0.0 0.0 667 37 9 71786.9 2 1 1 67734.79]
 [0.0 0.0 1.0 673 65 0 0.0 1 1 1 85733.33]
 [0.0 0.0 1.0 724 31 5 0.0 1 1 0 134889.95]
 [0.0 0.0 0.0 731 38 10 123711.73 2 1 0 171340.68]
 [0.0 1.0 0.0 484 39 5 0.0 2 1 1 175224.12]]


In [9]:
print(X_test[0:10, :])

[[1.0 0.0 0.0 597 35 8 131101.04 1 1 1 192852.67]
 [0.0 0.0 0.0 523 40 2 102967.41 1 1 0 128702.1]
 [0.0 1.0 0.0 706 42 8 95386.82 1 1 1 75732.25]
 [0.0 0.0 1.0 788 32 4 112079.58 1 0 0 89368.59]
 [1.0 0.0 1.0 706 38 5 163034.82 2 1 1 135662.17]
 [0.0 1.0 0.0 670 57 3 175575.95 2 1 0 99061.75]
 [0.0 1.0 1.0 590 34 0 65812.35 2 0 1 160346.3]
 [0.0 1.0 0.0 636 29 6 157576.47 2 1 1 101102.39]
 [0.0 0.0 0.0 598 64 9 0.0 1 0 1 13181.37]
 [0.0 0.0 0.0 456 63 1 165350.61 2 0 0 140758.07]]


In [10]:
print(y_train)

[1 0 0 ... 0 0 1]


In [11]:
print(y_test)

[0 1 0 ... 0 0 0]


# **Feature Scaling Using RobustScaler**

Note: I explained Feature Scaling earlier.

If you want to understand when and why to use feature scaling, you can check it here: [Go to  When is Feature Scaling Necessary in Regression Models? Section](https://github.com/Hazem1695/ml-concept-briefs)

In [12]:
from sklearn.preprocessing import RobustScaler
rs = RobustScaler()
X_train[:, [3,4,5,6,7,10]] = rs.fit_transform(X_train[:, [3,4,5,6,7,10]])
X_test[:, [3,4,5,6,7,10]] = rs.transform(X_test[:, [3,4,5,6,7,10]])

In [13]:
print(X_train[0:10, :])

[[1.0 0.0 0.0 -0.5367647058823529 0.16666666666666666 0.0
  0.16899040695224388 2.0 0 0 -0.9600941054381197]
 [0.0 0.0 0.0 0.7205882352941176 -0.4166666666666667 0.0
  -0.7565571396692488 1.0 1 0 -0.04540173752434738]
 [0.0 1.0 0.0 0.5661764705882353 -0.25 0.8 -0.33790089213121244 1.0 1 1
  -0.5828800244555142]
 [0.0 1.0 1.0 0.27205882352941174 0.08333333333333333 0.0
  -0.1668629561127746 0.0 1 1 -0.9313542915955845]
 [0.0 0.0 1.0 -0.34558823529411764 1.25 0.4 -0.7565571396692488 1.0 1 1
  0.7527444766300658]
 [0.0 0.0 0.0 0.11029411764705882 0.0 0.8 -0.19269115972501885 1.0 1 1
  -0.3294121290802671]
 [0.0 0.0 1.0 0.15441176470588236 2.3333333333333335 -1.0
  -0.7565571396692488 0.0 1 1 -0.14604121142311122]
 [0.0 0.0 1.0 0.5294117647058824 -0.5 0.0 -0.7565571396692488 0.0 1 0
  0.3547713277230695]
 [0.0 0.0 0.0 0.5808823529411765 0.08333333333333333 1.0
  0.21516382706766266 1.0 1 0 0.7261349944078792]
 [0.0 1.0 0.0 -1.2352941176470589 0.16666666666666666 0.0
  -0.7565571396692488 1

In [14]:
print(X_test[0:10, :])

[[1.0 0.0 0.0 -0.40441176470588236 -0.16666666666666666 0.6
  0.27320478613208393 0.0 1 1 0.9453012905036494]
 [0.0 0.0 0.0 -0.9485294117647058 0.25 -0.6 0.052223007432371495 0.0 1 0
  0.2917288957910999]
 [0.0 1.0 0.0 0.39705882352941174 0.4166666666666667 0.6
  -0.007320402422426916 0.0 1 1 -0.24793321028316476]
 [0.0 0.0 1.0 1.0 -0.4166666666666667 -0.2 0.12379654893462903 0.0 0 0
  -0.10900482057572297]
 [1.0 0.0 1.0 0.39705882352941174 0.08333333333333333 0.0
  0.5240356780221564 1.0 1 1 0.36263878197398147]
 [0.0 1.0 0.0 0.1323529411764706 1.6666666666666667 -0.4
  0.622542739297375 1.0 1 0 -0.010249941335649226]
 [0.0 1.0 1.0 -0.45588235294117646 -0.25 -1.0 -0.23961957550109766 1.0 0
  1 0.6141231562437014]
 [0.0 1.0 0.0 -0.11764705882352941 -0.6666666666666666 0.2
  0.4811618685718447 1.0 1 1 0.01054030217342975]
 [0.0 0.0 0.0 -0.39705882352941174 2.25 0.8 -0.7565571396692488 0.0 0 1
  -0.8852078039925787]
 [0.0 0.0 0.0 -1.4411764705882353 2.1666666666666665 -0.8
  0.5422255594

# **Handling Class Imbalance using SMOTE**

### Difference Between Handling Class Imbalance Methods

| Method        | Action                             |
| ------------- | ---------------------------------- |
| Class Weights | changes learning importance        |
| SMOTE         | creates synthetic minority samples |
| Undersampling | removes majority samples           |


Note: I explained handling imbalanced data earlier.

If you want to understand when and why to use techniques such as Class Weights, SMOTE, and Undersampling, you can check:
[Handling Imbalanced Data in Machine Learning](https://github.com/Hazem1695/ml-concept-briefs)


In [15]:
!pip install imbalanced-learn

In [16]:
from imblearn.over_sampling import SMOTE
smote = SMOTE(random_state=0)
X_train_smote, y_train_smote = smote.fit_resample(X_train,y_train)

In [17]:
print(X_train_smote[0:10, :])

[[ 1.          0.          0.         -0.53676471  0.16666667  0.
   0.16899041  2.          0.          0.         -0.96009411]
 [ 0.          0.          0.          0.72058824 -0.41666667  0.
  -0.75655714  1.          1.          0.         -0.04540174]
 [ 0.          1.          0.          0.56617647 -0.25        0.8
  -0.33790089  1.          1.          1.         -0.58288002]
 [ 0.          1.          1.          0.27205882  0.08333333  0.
  -0.16686296  0.          1.          1.         -0.93135429]
 [ 0.          0.          1.         -0.34558824  1.25        0.4
  -0.75655714  1.          1.          1.          0.75274448]
 [ 0.          0.          0.          0.11029412  0.          0.8
  -0.19269116  1.          1.          1.         -0.32941213]
 [ 0.          0.          1.          0.15441176  2.33333333 -1.
  -0.75655714  0.          1.          1.         -0.14604121]
 [ 0.          0.          1.          0.52941176 -0.5         0.
  -0.75655714  0.          1

## Data Dimensions Before SMOTE

In [18]:
 print(f"Rows: {X_train.shape[0]}, Columns: {X_train.shape[1]}")

Rows: 7500, Columns: 11


In [19]:
print(f"Rows: {y_train.shape[0]}")

Rows: 7500


## Data Dimensions After SMOTE

In [20]:
 print(f"Rows: {X_train_smote.shape[0]}, Columns: {X_train_smote.shape[1]}")

Rows: 11944, Columns: 11


In [21]:
print(f"Rows: {y_train_smote.shape[0]}")

Rows: 11944


# **Building Naive Bayes Classification Model**

## Training the Building Naive Bayes Classification model on the Training set

In [22]:
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train_smote, y_train_smote)

GaussianNB()

# **Preticting the Test set results**

In [23]:
y_pred = classifier.predict(X_test)

print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1)[0:100,:])

[[1 0]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [1 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [1 1]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [1 0]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [0 0]
 [0 1]
 [1 1]
 [1 1]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [0 0]
 [1 1]
 [0 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [1 0]
 [1 0]
 [1 0]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 0]]


# **Evaluating the Model Performance**

In [24]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report, roc_auc_score
from sklearn.model_selection import cross_val_score

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nAccuracy Score:")
print(accuracy_score(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nROC-AUC Score:")
print(roc_auc_score(y_test, y_pred))


accuracies = cross_val_score(estimator=classifier, X=X_train, y=y_train, cv=10)

print("\nMean Accuracy:")
print(accuracies.mean())

print("\nStandard Deviation:")
print(accuracies.std())


Confusion Matrix:
[[1429  562]
 [ 148  361]]

Accuracy Score:
0.716

Classification Report:
              precision    recall  f1-score   support

           0       0.91      0.72      0.80      1991
           1       0.39      0.71      0.50       509

    accuracy                           0.72      2500
   macro avg       0.65      0.71      0.65      2500
weighted avg       0.80      0.72      0.74      2500


ROC-AUC Score:
0.7134817878883266

Mean Accuracy:
0.8214666666666666

Standard Deviation:
0.009284634857895295


# Gaussian Naive Bayes Model Comparison and Evaluation

This experiment focused on evaluating different configurations of Gaussian Naive Bayes models for an imbalanced classification problem. The comparison investigated the impact of:

* Synthetic Minority Oversampling Technique
* Prior probability adjustment
* Feature scaling strategies
* One-hot encoding strategies
* Minority-class detection capability
* Model stability and predictive balance

The experiments aimed to determine how preprocessing and imbalance-handling strategies influence the performance of Gaussian Naive Bayes classifiers.

---

# 1. Gaussian Naive Bayes Without Synthetic Minority Oversampling Technique Using StandardScaler

The baseline Gaussian Naive Bayes model achieved an overall accuracy score of `79.64%`. However, the confusion matrix revealed that the model predicted every sample as the majority class and completely failed to classify minority-class samples.

As a result:

* Minority-class precision = `0.00`
* Minority-class recall = `0.00`
* Minority-class F1-score = `0.00`
* Receiver Operating Characteristic - Area Under Curve = `0.50`

A Receiver Operating Characteristic - Area Under Curve score of `0.50` indicates no effective class separation capability beyond random guessing.

The warning messages generated during evaluation confirmed that the model never predicted minority-class labels.

This experiment demonstrates that Gaussian Naive Bayes struggles significantly with imbalanced datasets when no imbalance-handling strategy is applied.

---

# 2. Gaussian Naive Bayes With Balanced Priors Using StandardScaler

After adjusting the class prior probabilities to `[0.5, 0.5]`, the model demonstrated a major improvement in minority-class detection.

The minority-class recall increased dramatically to `0.68`, while the Receiver Operating Characteristic - Area Under Curve score improved to `0.623`.

This improvement indicates that manually balancing class priors encouraged the model to become more sensitive to minority-class samples.

However, this configuration also produced:

* Lower overall accuracy = `58.76%`
* High number of false positives
* Reduced majority-class performance

Although the model became more balanced, the trade-off between recall and precision remained significant.

This experiment highlights the importance of prior probability adjustment in probabilistic classifiers when dealing with imbalanced datasets.

---

# 3. Gaussian Naive Bayes With Synthetic Minority Oversampling Technique Using StandardScaler

Applying Synthetic Minority Oversampling Technique improved minority-class recall further to `0.76`, which was one of the highest recall values across all Gaussian Naive Bayes experiments.

However, the model suffered from severe performance imbalance:

* Accuracy dropped to `48.80%`
* Minority-class precision remained very low at `0.25`
* Receiver Operating Characteristic - Area Under Curve remained relatively weak at `0.589`

The confusion matrix showed a very large number of false positives, meaning the model over-predicted the minority class.

This experiment demonstrates that while Synthetic Minority Oversampling Technique improved sensitivity toward minority samples, it also introduced instability and excessive false-positive predictions in Gaussian Naive Bayes.

---

# 4. Gaussian Naive Bayes With Synthetic Minority Oversampling Technique Using RobustScaler

Using RobustScaler instead of StandardScaler produced very similar behavior.

The model achieved:

* Accuracy score = `47.76%`
* Minority recall = `0.74`
* Minority F1-score = `0.37`
* Receiver Operating Characteristic - Area Under Curve = `0.576`

Although RobustScaler slightly reduced the minority recall compared to StandardScaler, the overall performance remained highly unstable.

The model still generated a large number of false positives and struggled to maintain balanced predictive capability.

This experiment suggests that changing the scaling strategy alone was insufficient to stabilize Gaussian Naive Bayes performance after applying Synthetic Minority Oversampling Technique.

---

# 5. Gaussian Naive Bayes With Synthetic Minority Oversampling Technique, RobustScaler, and One-Hot Encoding With Drop First

This experiment produced the strongest overall performance among all Gaussian Naive Bayes configurations.

The model achieved:

* Accuracy score = `71.60%`
* Minority recall = `0.71`
* Minority F1-score = `0.50`
* Receiver Operating Characteristic - Area Under Curve = `0.713`

Compared to previous experiments, the addition of one-hot encoding with dropping the first category significantly improved model stability and overall classification balance.

The confusion matrix showed a substantial reduction in false positives while maintaining strong minority-class recall.

This experiment demonstrates that feature representation and encoding strategies strongly influence Gaussian Naive Bayes performance.

---

# 6. Gaussian Naive Bayes With Synthetic Minority Oversampling Technique, StandardScaler, and One-Hot Encoding With Drop First

This experiment also produced strong and balanced performance.

The model achieved:

* Accuracy score = `71.56%`
* Minority recall = `0.69`
* Minority F1-score = `0.50`
* Receiver Operating Characteristic - Area Under Curve = `0.707`

The results remained extremely close to the RobustScaler configuration.

This experiment confirms that combining:

* Synthetic Minority Oversampling Technique
* Feature scaling
* One-hot encoding optimization

significantly improves Gaussian Naive Bayes performance in imbalanced classification problems.

---

# Performance Comparison Table

| Experiment                                                                                                         | Accuracy | Minority Precision | Minority Recall | Minority F1-Score | Receiver Operating Characteristic - Area Under Curve |
| ------------------------------------------------------------------------------------------------------------------ | -------- | ------------------ | --------------- | ----------------- | ---------------------------------------------------- |
| Gaussian Naive Bayes Without Synthetic Minority Oversampling Technique + StandardScaler                            | 0.796    | 0.00               | 0.00            | 0.00              | 0.500                                                |
| Gaussian Naive Bayes With Balanced Priors + StandardScaler                                                         | 0.588    | 0.29               | 0.68            | 0.40              | 0.623                                                |
| Gaussian Naive Bayes With Synthetic Minority Oversampling Technique + StandardScaler                               | 0.488    | 0.25               | 0.76            | 0.38              | 0.589                                                |
| Gaussian Naive Bayes With Synthetic Minority Oversampling Technique + RobustScaler                                 | 0.478    | 0.24               | 0.74            | 0.37              | 0.576                                                |
| Gaussian Naive Bayes With Synthetic Minority Oversampling Technique + RobustScaler + One-Hot Encoding Drop First   | 0.716    | 0.39               | 0.71            | 0.50              | 0.713                                                |
| Gaussian Naive Bayes With Synthetic Minority Oversampling Technique + StandardScaler + One-Hot Encoding Drop First | 0.716    | 0.39               | 0.69            | 0.50              | 0.707                                                |

---

# Ranking of Experiments

| Rank | Experiment                                                                                                         | Main Strength                                                                         |
| ---- | ------------------------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------- |
| 1    | Gaussian Naive Bayes With Synthetic Minority Oversampling Technique + RobustScaler + One-Hot Encoding Drop First   | Best overall balance and highest Receiver Operating Characteristic - Area Under Curve |
| 2    | Gaussian Naive Bayes With Synthetic Minority Oversampling Technique + StandardScaler + One-Hot Encoding Drop First | Strong balanced performance and stability                                             |
| 3    | Gaussian Naive Bayes With Balanced Priors + StandardScaler                                                         | Strong recall improvement using prior adjustment                                      |
| 4    | Gaussian Naive Bayes With Synthetic Minority Oversampling Technique + StandardScaler                               | Highest minority recall                                                               |
| 5    | Gaussian Naive Bayes With Synthetic Minority Oversampling Technique + RobustScaler                                 | Similar recall but weaker stability                                                   |
| 6    | Gaussian Naive Bayes Without Synthetic Minority Oversampling Technique + StandardScaler                            | Failed to detect minority-class samples                                               |

---

# Key Insights

## Accuracy Alone Is Misleading in Imbalanced Classification Problems

The baseline model achieved high accuracy while completely failing to detect minority-class samples. This confirms that accuracy alone is not a sufficient metric for evaluating imbalanced datasets.

---

## Gaussian Naive Bayes Is Highly Sensitive to Imbalance

Without imbalance-handling strategies, Gaussian Naive Bayes strongly favored the majority class and ignored minority samples entirely.

This demonstrates the importance of preprocessing and balancing techniques when using probabilistic classifiers.

---

## Prior Probability Adjustment Improved Minority-Class Detection

Balancing class priors significantly improved minority recall and Receiver Operating Characteristic - Area Under Curve performance.

This shows that prior probability tuning can partially compensate for class imbalance in Gaussian Naive Bayes models.

---

## Synthetic Minority Oversampling Technique Increased Recall but Introduced False Positives

Applying Synthetic Minority Oversampling Technique greatly improved minority recall but also generated excessive false-positive predictions in several experiments.

This highlights the trade-off between sensitivity and precision in imbalanced classification problems.

---

## One-Hot Encoding Optimization Significantly Improved Performance

The strongest results were achieved after combining:

* Synthetic Minority Oversampling Technique
* Feature scaling
* One-hot encoding with dropping the first category

This indicates that feature representation quality strongly affects Gaussian Naive Bayes performance.

---

## RobustScaler and StandardScaler Produced Similar Final Performance

Both scaling strategies achieved highly competitive results when combined with optimized preprocessing pipelines.

This suggests that preprocessing consistency was more important than the specific scaling method itself.

---

# Conclusion

This experimentation demonstrated that Gaussian Naive Bayes performance is strongly influenced by imbalance-handling techniques, preprocessing strategies, and feature representation methods.

The experiments showed that:

* Gaussian Naive Bayes fails to detect minority-class samples without imbalance handling
* Prior probability adjustment improves recall performance
* Synthetic Minority Oversampling Technique increases minority sensitivity but may introduce false positives
* One-hot encoding optimization significantly improves model stability and predictive balance
* Combining preprocessing strategies produces substantially stronger classification performance

The strongest overall configuration was Gaussian Naive Bayes with Synthetic Minority Oversampling Technique, RobustScaler, and one-hot encoding with dropping the first category, as it achieved the best balance between minority recall, F1-score, Receiver Operating Characteristic - Area Under Curve, and overall predictive stability.

These experiments highlight the importance of systematic preprocessing experimentation and careful evaluation when building probabilistic machine learning pipelines for imbalanced datasets.
